util/exercise/src/cmd/generate.rs

```diff
@@ -78,50 +46,11 @@ fn generate_tests_from_canonical_data(
 ) -> Result<()> {
     exercise::update_cargo_toml_version(exercise_name, canonical_data)?;
 
-    let tests_content = exercise::get_tests_content(exercise_name)?;
-
-    let updated_tests_content = format!(
-        "//! Tests for {exercise_name} \n\
-        //! \n\
-        //! Generated by [utility][utility] using [canonical data][canonical_data]\n\
-        //! \n\
-        //! [utility]: https://github.com/exercism/rust/tree/master/util/exercise\n\
-        //! [canonical_data]: https://raw.githubusercontent.com/exercism/problem-specifications/master/exercises/{exercise_name}/canonical-data.json\n\
-        \n\
-        {} \n\
-        ",
-        tests_content,
-        exercise_name=exercise_name,
-    );
-
-    fs::write(&tests_path, updated_tests_content)?;
-
-    let mut property_functions: HashMap<String, String> = HashMap::new();
-
-    let mut test_functions: Vec<String> = Vec::new();
-
-    for case in &canonical_data.cases {
-        parse_case(
-            case,
-            &mut property_functions,
-            &mut test_functions,
-            use_maplit,
-        )?;
-    }
-
-    if !test_functions.is_empty() {
-        let first_test_function = test_functions.remove(0).replace("#[ignore]\n", "");
-
-        test_functions.insert(0, first_test_function);
-    }
-
-    let mut tests_file = OpenOptions::new().append(true).open(&tests_path)?;
-
-    for property_body in property_functions.values() {
-        tests_file.write_all(property_body.as_bytes())?;
-    }
+    let mut context = Context::from_serialize(canonical_data)?;
```

Comment:
While it's cool that you've managed to extract the set of properties within the template expansion, I do feel like the test file is much easier to read if the properties list comes first. It should be pretty straightforward to recursively traverse `canonical_data` at this point, discovering the proper set of properties, and then inject the properties list into the context.

Reply:
I agree on the ordering. To be honest, I held off on implementing it because I expect that unpacking the inputs correctly will require grouping the test cases by property, to make sure all possible input formats are covered (for example, the rest_api exercise has a variable number of inputs).

But, we might as well do it now. :smile:

Reply:
It was indeed pretty straightforward to implement in the end (see [afc5929](https://github.com/exercism/rust/pull/908/commits/afc592977fe4afc504b9451f00d5c95a44b81cfb)).